### ISSUE TITLE
Bug may lead to memory leak

### ISSUE THREAD

**Issue by sjy1203** [illidanlab/T-LSTM#2]
### Bug Code
`c_test, y_pred_test, y_test, logits_test, labels_test  = sess.run(lstm.get_cost_acc(), feed_dict={lstm.input: batch_xs, lstm.labels: batch_ys, lstm.time: batch_ts, lstm.keep_prob: test_dropout_prob})
            `

### Right Code
`c_test, y_pred_test, y_test, logits_test, labels_test  = sess.run([cross_entropy, y_pred, y, logits, labels], feed_dict={lstm.input: batch_xs, lstm.labels: batch_ys, lstm.time: batch_ts, lstm.keep_prob: test_dropout_prob})`

### Explaination
 lstm.get_cost_acc() will add new tf.op to the tensorflow graph. In my test, this program will consume my server's all 30G memory at about 40th epoch. And It will slow down the test procedure.

**Comment by baytasin** [comment]
Thank you very much for sharing your feedback. In that setting, if you call cross entropy in the test phase, it may not reflect the test performance correctly. That was why I called lstm.get_cost_acc() in the set. But I updated the main.py and TLSTM.py into a newer version. You can try them out.

**Comment by sjy1203** [comment]
Thank you for your prompt reply！Sorry, I am not very familiar with tensorflow. Can you explain why 'cross entropy' in the test phase may not reflect the test performance correctly?(Or give some useful links) 
I read the newer version which is more readable. Thank you! But there is a problem I want to plot the learning curve which means  I need to record the test result every epoch.

**Comment by baytasin** [comment]
No problem. If we use sess.run (cross_entropy) directly, we cannot pass the test data. You might need to use eval. In my implementation, I prefer to use classes and the current version worked well for me. Please let me know if you have the same memory issue. Can you please explain what is the problem with plotting the learning curve?

**Comment by sjy1203** [comment]
Thank you for your explanation. Now I just use the following code to get the results in test phase. It can actually pass the test data like you did in train phase. 
Let's talk about the problem with plotting the learning curve. For example, I want to train my model for about 50 epochs, for every epoch I want to record my AUC or F1 score in test data. What your code does now in test phase is 'lstm.get_cost_acc()', it will add new tf.op to the tf.graph and lead to memory leak problem. The memory of your server will be exhausted. So I still think you just need to replace 'lstm.get_cost_acc()' with '[cross_entropy, y_pred, y, logits, labels]' which you have already generated before the train and test phase.


### My Code
` c_test, y_pred_test, y_test, logits_test, labels_test = sess.run([cross_entropy, y_pred, y, logits, labels], feed_dict={lstm.input: batch_xs, lstm.labels: batch_ys, lstm.time: batch_ts, lstm.keep_prob: test_dropout_prob})`

**Comment by 2g-XzenG** [comment]
@sjy1203 
Hi sjy,

thanks for pointing out this issue, I also run into the memory issue, I notice one way to solve this might be:
```
acc = lstm.get_cost_acc()

c_test, y_pred_test, y_test, logits_test, labels_test = sess.run(acc, feed_dict={lstm.input: batch_xs, lstm.labels: batch_ys, lstm.time: batch_ts, lstm.keep_prob: test_dropout_prob})
```

Thanks

**Comment by baytasin** [comment]
Hi @1230pitchanqw , thank you for your feedback. Could you please share more details about your experimental setting? I would like to reproduce the memory leak problem and I will update the code accordingly. Thank you.
